trieste/utils/pareto.py

```diff
@@ -0,0 +1,309 @@
+        self.lb = lb
+        self.ub = ub
+
+    def append(self, lb: tf.Tensor, ub: tf.Tensor):
+        """
+        Add new bounded volumes.
+        :param lb: the lowerbounds of the volumes
+        :param ub: the upperbounds of the volumes
+        """
+        self.lb = tf.concat([self.lb, lb], axis=0)
+        self.ub = tf.concat([self.ub, ub], axis=0)
+
+    def clear(self):
+        """
+        Clears all stored bounded volumes
+        """
+        dtype = self.lb.dtype
+        outdim = self.lb.shape[1]
+        self.lb = tf.zeros([0, outdim], dtype=dtype)
+        self.ub = tf.zeros([0, outdim], dtype=dtype)
+
+    def size(self) -> tf.Tensor:
+        """
+        :return: volume of each bounded volume
+        """
+        return tf.math.reduce_prod(self.ub - self.lb, axis=0, keepdims=True)
+
+
+class Pareto:
+    def __init__(self, datasets: Mapping[str, Dataset], threshold: tf.Tensor = 0):
+        """
+        Construct a Pareto set.
+        Stores a Pareto set and calculates the cell bounds covering the non-dominated region.
+        The latter is needed for certain multiobjective acquisition functions.
+        :param datasets: Mapping of Dataset and its coresponding string key(s)
+        :param threshold: approximation threshold for the generic divide and conquer strategy
+            (default 0: exact calculation)
+        """
+        self.threshold = threshold
+        self.data = to_Dataset(datasets)
+
+        # Setup data structures
+        self.bounds = BoundedVolumes(
+            tf.zeros([0, self.data.observations.shape[1]], dtype=tf.int32),
+            tf.zeros([0, self.data.observations.shape[1]], dtype=tf.int32),
+        )
+
+        self.front = tf.zeros([0, self.data.observations.shape[1]])
+        # Initialize
+        self.update()
+
+    @staticmethod
+    def _is_test_required(smaller: tf.Tensor) -> tf.Tensor:
+        """
+        Tests if a point augments or dominates the Pareto set.
+        :param smaller: a boolean tf.Tensor storing test point < Pareto front
+        :return: True if the test point dominates or augments the Pareto front (boolean)
+        """
+        # if and only if the test point is at least in one dimension smaller
+        # for every point in the Pareto set
+        idx_dom_augm = tf.reduce_any(smaller, axis=1)
+        is_dom_augm = tf.reduce_all(idx_dom_augm)
+
+        return is_dom_augm
+
+    def _update_front(self) -> tf.Tensor:
+        """
+        Calculate the non-dominated set of points based on the latest data.
+        The stored Pareto set is sorted on the first objective in ascending order.
+        :return: boolean, whether the Pareto set has actually changed since the last iteration
+        """
+        current = self.front
+        pf, _ = non_dominated_sort(self.data)
+        self.front = tf.gather_nd(pf, tf.expand_dims(tf.argsort(pf[:, 0]), axis=1))
+        return not np.array_equal(current.numpy(), self.front.numpy())
```

Comment:
array equality is a tricky one. How do we handle points that are effectively equal but not exactly due to precision errors?